Renode sanity check

PiotrZierhoffer opened the pull request:
This pull requests adds Renode support for sanitycheck.

The tests on Renode are enabled for `m2gl025_miv` platform.

To run tests on Renode, you need to have it installed from a package.
Please download Renode from [the GitHub repository](https://github.com/renode/renode) and follow the [build instructions](https://renode.readthedocs.io/en/latest/advanced/building_from_sources.html).

To start tests, run `./scripts/sanitycheck --platform m2gl025_miv -v`

It is still work in progress. Renode hangs on failed tests, we need to solve closing the framework on timeout.

We are very interested in knowing if this approach (via cmake files and console analysis) is right and how should we move forward with enabling more platforms.

SebastianBoe commented:
The general approach wrt. the build system looks aligned with qemu, and is therefore OK. There are some code-technical improvements needed, but we can discuss those when this is no longer an RFC.

PiotrZierhoffer commented:
One question we'll need to address is how to enable Renode for platforms that already have an emulator - like `hifive1`, which can be started on QEMU (it has `EMU_PLATFORM` variable already set).

SebastianBoe commented:
> One question we'll need to address is how to enable Renode for platforms that already have an emulator - like hifive1, which can be started on QEMU (it has EMU_PLATFORM variable already set).

Hmm, perhaps we could make EMU_PLATFORM into a list. And use some prefixes in the custom target names to allow running either emulator. Or we could have the user choose an emulator through Kconfig/CMake.

nashif commented:
any instructions on building renode on fedora 29?

nashif commented:
seems to work fine, you need to fix your commit messages for this to go in.

wojciechowskidawid committed:
sanitycheck: Add Renode cmake configuration

By default Renode is searched for in the system PATH.
The PID is stored in renode.pid - the application must be killed with a
signal after a test.

With the XWT gui layer disabled all UART output is directed to the
stdout.

Signed-off-by: Dawid Wojciechowski <dwojciechowski@internships.antmicro.com>
Signed-off-by: Jan Kowalewski <jkowalewski@antmicro.com>
Signed-off-by: Piotr Zierhoffer <pzierhoffer@antmicro.com>

Commit:
sanitycheck: Add Renode support

The change adds an optional capability to kill process by PID in the
BinaryHandler. At the moment proc.terminate only closed the make
process, also closing the stdin/stdout, but this is impossible to detect
in .NET.

Signed-off-by: Dawid Wojciechowski <dwojciechowski@internships.antmicro.com>
Signed-off-by: Jan Kowalewski <jkowalewski@antmicro.com>
Signed-off-by: Piotr Zierhoffer <pzierhoffer@antmicro.com>

wojciechowskidawid committed:
sanitycheck: Enable Renode tests on m2gl025_miv

The test suite will use the m2gl025_miv.resc Renode script to load the
platform.

Signed-off-by: Dawid Wojciechowski <dwojciechowski@internships.antmicro.com>
Signed-off-by: Jan Kowalewski <jkowalewski@antmicro.com>
Signed-off-by: Piotr Zierhoffer <pzierhoffer@antmicro.com>

PiotrZierhoffer commented:
The previous push from ddf408a to 2752590 was a mistake.

Current state uses the BinaryHandler, improved with the capability of sending a signal directly to the simulator.

112 out of 113 tests pass. The one that does not is also failing on hardware, we are still investigating why that happens. The failing test is `tests/kernel/mem_pool/mem_pool_api/kernel.memory_pool`

nashif commented:
```
Traceback (most recent call last):
  File "/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/scripts/sanitycheck", line 3082, in <module>
    main()
  File "/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/scripts/sanitycheck", line 3007, in main
    ts.instances)
  File "/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/scripts/sanitycheck", line 2137, in execute
    mg.add_test_instance(i, options.extra_args)
  File "/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/scripts/sanitycheck", line 1106, in add_test_instance
    self.add_goal(ti, type, args)
  File "/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/scripts/sanitycheck", line 1038, in add_goal
    handler = BinaryHandler(instance)
  File "/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/scripts/sanitycheck", line 349, in __init__
    if instance.pid_file != None:
AttributeError: 'TestInstance' object has no attribute 'pid_file'
```

nashif commented:
> If I understand correctly, to have the first command issue tests for m2gl025_miv I have to add `default: true` to the relevant yml. Is that something we want to have? I didn't find any guidelines for that.

we can make it default once we have everything in place, right now lets leave it as is and call sanitycheck with the platform.

PiotrZierhoffer commented:
I am currently preparing the Docker image, as we discussed earlier. I will post a link to the relevant pull request when I'm done.

PiotrZierhoffer commented:
The Docker-related pull request is here: https://github.com/zephyrproject-rtos/docker-image/pull/6

All Mi-V (113) tests pass properly. `./scripts/sanitycheck` started in Docker also works fine.

PiotrZierhoffer commented on boards/riscv32/m2gl025_miv/m2gl025_miv.resc:
```diff
@@ -0,0 +1,16 @@
+:name: Mi-V
```
done
